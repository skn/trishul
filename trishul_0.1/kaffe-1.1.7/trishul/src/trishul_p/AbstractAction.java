package trishul_p;

public abstract class AbstractAction
extends Action
{

	/* These trishul hook methods are intentionally not integrated with the Engine class, as this
	 * allows the this pointers to be checked
	 */
	final protected native boolean __TRISHUL_MATCHES 	(TrishulActionPattern pattern, Action action);


	/**
	 * Check whether parameter can be matched by this abstract action.
	 */
	public abstract boolean matches (Action a);
}
