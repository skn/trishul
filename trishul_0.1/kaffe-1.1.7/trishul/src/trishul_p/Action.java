package trishul_p;

public abstract class Action
extends Throwable
{
	public abstract Object getThisPointer ();

	public abstract int	   getParameterCount ();
	public abstract Object getActualParameter (int index);
	public abstract Class  getFormalParameter (int index);

	public String getContext ()
	{
		return getContext (0);
	}

	public String getContext (int depth)
	{
		StackTraceElement stack = getStackTrace ()[depth];
		return stack.getClassName () + "." + stack.getMethodName ();
	}
}

