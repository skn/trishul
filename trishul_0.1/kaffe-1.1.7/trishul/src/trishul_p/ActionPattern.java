package trishul_p;

import japa.parser.ast.expr.Expression;
import japa.parser.ast.type.Type;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;

public class ActionPattern
{
	public static class FormalParameter
	{
		/// null means any type
		private Type	type;

		/// name of local variable created to hold actual value
		private String	name;

		/// null means match any (or no) taint.
		private TaintPattern taint;

		/**
		 *
		 */
		public FormalParameter (Type type, String name, TaintPattern taint)
		{
			this.type	= type;
			this.name	= name;
			this.taint	= taint;
		}

		public Type getType ()
		{
			return type;
		}

		public boolean hasType ()
		{
			return (type != null);
		}

		public String getName ()
		{
			return name;
		}

		public boolean hasName ()
		{
			return (name != null);
		}

		public TaintPattern getTaint ()
		{
			return taint;
		}

		public boolean hasTaint ()
		{
			return (taint != null);
		}
	}

	private boolean							isAbstractAction;
	private boolean 						isFinal;
	private List<FormalParameter>			formals = new LinkedList<FormalParameter> ();
	private boolean							ellipses;
	private TaintPattern					ellipsesTaint;
	private Type							returnType;
	private String 							packageName;
	private String 							className;
	private String 							methodName;
	private TaintPattern 					thisTaint;
	private TaintPattern 					contextTaint;

	public ActionPattern ()
	{
	}

	public void setAbstractAction (boolean isAbstractAction)
	{
		this.isAbstractAction = isAbstractAction;
	}

	public boolean isAbstractAction ()
	{
		return isAbstractAction;
	}

	public String getAbstractActionName ()
	{
		return packageName + "." + className + "." + methodName;
	}

	public void setFinal (boolean isFinal)
	{
		this.isFinal = isFinal;
	}

	public boolean isFinal ()
	{
		return isFinal;
	}

	public void setMethod (String packageName, String className, String methodName, TaintPattern thisTaint)
	{
		this.packageName	= packageName;
		this.className		= className;
		this.methodName		= methodName;
		this.thisTaint		= thisTaint;
	}

	public String getPackageName ()
	{
		return packageName;
	}

	public boolean hasPackageName ()
	{
		return (packageName != null);
	}

	public String getClassName ()
	{
		return className;
	}

	public boolean hasClassName ()
	{
		return (className != null);
	}

	public String getMethodName ()
	{
		return methodName;
	}

	public boolean hasMethodName ()
	{
		return (methodName != null);
	}

	public TaintPattern getThisTaint ()
	{
		return thisTaint;
	}

	public boolean hasThisTaint ()
	{
		return (thisTaint != null);
	}

	public void setContextTaint (TaintPattern contextTaint)
	{
		this.contextTaint = contextTaint;
	}

	public TaintPattern getContextTaint ()
	{
		return contextTaint;
	}

	public boolean hasContextTaint ()
	{
		return (contextTaint != null);
	}

	public void addFormalParameter (FormalParameter formal)
	{
		formals.add (formal);
	}

	public Collection<FormalParameter> formalParameters ()
	{
		return formals;
	}

	public boolean hasFormalParameters ()
	{
		return !formals.isEmpty ();
	}

	public void allowAnyFormalParameter (TaintPattern taint)
	{
		ellipses 		= true;
		ellipsesTaint 	= taint;
	}

	public boolean isAnyParameterAllowed ()
	{
		return ellipses;
	}

	public TaintPattern getAnyParameterTaint ()
	{
		return ellipsesTaint;
	}

	public boolean hasAnyParameterTaint ()
	{
		return (ellipsesTaint != null);
	}

	public void setReturnType (Type type)
	{
		this.returnType = type;
	}

	public Type getReturnType ()
	{
		return returnType;
	}

	public boolean hasReturnType ()
	{
		return (returnType != null);
	}

	public String toString ()
	{
		String s = "<";

		// Modifiers
		if (isFinal) s += "final";
		if (isAbstractAction)
		{
			if (isFinal) s += ", ";
			s += "final";
		}
		if (isFinal || isAbstractAction)
		{
			s += " ";
		}

		// Return type
		if (returnType == null)
		{
			s += "*";
		}
		else
		{
			s += returnType.toString ();
		}

		// Method name
		s += " ";
		if (packageName == null)
			s += "*";
		else
			s += packageName;
		s += ".";
		if (className == null)
			s += "*";
		else
			s += className;
		if (thisTaint != null)
			s += thisTaint.toString ();
		s += ".";
		if (methodName == null)
			s += "*";
		else
			s += methodName;

		// Formal parameters
		s += "(";
		boolean first = true;

		for (FormalParameter fp: formals)
		{
			if (!first) s += ", ";

			if (fp.type == null)
				s += "*";
			else
				s += fp.type.toString ();

			if (fp.name != null)
				s += " " + fp.name;

			if (fp.taint != null)
				s += fp.taint.toString ();

			first = false;
		}

		if (ellipses)
		{
			if (!first) s += ", ";
			s += "..";
			if (ellipsesTaint != null)
				s += ellipsesTaint.toString ();
		}
		s += ")";

		if (contextTaint != null)
			s += contextTaint.toString ();

		s += ">";
		return s;
	}


}

