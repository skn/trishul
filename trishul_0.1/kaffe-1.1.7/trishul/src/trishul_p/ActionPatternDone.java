package trishul_p;

public class ActionPatternDone
extends ActionPattern
{
	public ActionPatternDone ()
	{
	}

	public String toString ()
	{
		return "done";
	}
}