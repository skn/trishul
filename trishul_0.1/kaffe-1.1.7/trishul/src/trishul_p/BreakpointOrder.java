package trishul_p;

/**
 * Causes a trap into the debugger. Purely for debugging purposes.
 */
public class BreakpointOrder
extends Order
{
 	public BreakpointOrder ()
  	{
  	}

	public int getType ()
	{
		return TYPE_BREAKPOINT;
	}
}
