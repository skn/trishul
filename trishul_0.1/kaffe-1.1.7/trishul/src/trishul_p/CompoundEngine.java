package trishul_p;
import java.util.LinkedList;
import java.util.List;

public class CompoundEngine
extends Engine
{
	private List<Engine>	engines = new LinkedList<Engine> ();

	public CompoundEngine ()
	{
	}

	/**
	 * Adds an engine to the end of the list.
	 */
	public void addEngine (Engine newEngine)
	{
		engines.add (newEngine);
	}

	/**
	 *
	 */
	public void addEngine (Engine after, Engine newEngine)
	{
		int index = engines.indexOf (after);
		if (index < 0)
			throw new IllegalArgumentException ("Engine not found");
		engines.add (index + 1, newEngine);
	}

	public void removeEngine (Engine remove)
	{
		// This is done using a copy, to avoid a ConcurrentModificationException
		int oldSize = engines.size ();
		List<Engine>	engines = new LinkedList<Engine> ();
		engines.addAll (this.engines);
		engines.remove (remove);
		if (oldSize == engines.size ())
			throw new IllegalArgumentException ("Engine not found");
		this.engines = engines;
	}

	public Order query (Action a)
	{
		Order 			result 		= null;
		List<Order> 	compound 	= new LinkedList<Order> ();

	outOfLoop:
		for (Engine engine: engines)
		{
			Order partial = engine.doQuery (a);
			if (partial != null)
			{
				result = partial;
				compound.add (partial);

				switch (partial.getType ())
				{
				case Order.TYPE_HALT:
				case Order.TYPE_REPLACE:
				case Order.TYPE_SUPPRESS:
				case Order.TYPE_EXCEPTION:
					break outOfLoop;
				}
			}
		}

		if (compound.size () > 1)
		{
			result = new CompoundOrder (compound.toArray (new Order[0]));
		}

		return result;
	}

}
