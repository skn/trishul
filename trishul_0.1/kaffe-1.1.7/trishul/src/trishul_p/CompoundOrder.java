package trishul_p;

/**
 *
 */
public class CompoundOrder
extends Order
{
	private Order[] orders;

	public CompoundOrder (Order[] orders)
	{
		this.orders = orders;
	}

	public Order[] getOrders ()
	{
		return orders;
	}

	public int getType ()
	{
		return TYPE_COMPOUND;
	}
}
