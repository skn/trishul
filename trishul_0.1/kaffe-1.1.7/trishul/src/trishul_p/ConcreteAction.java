package trishul_p;

import java.util.StringTokenizer;

public class ConcreteAction
extends Action
{
	private String 		packageName;
	private String 		methodName;
	private String 		className;
	private String 		signature;
	private Object 		thisPointer;
	private int			methodHandle;
	private int			actualParameterHandle;
	private Object[] 	actualParameters;
	private Class[] 	formalParameters;

	/**
	 * Private constructor for kaffe only.
	 */
  	private ConcreteAction (Object thisPointer,
  							String packageName, String className, String methodName, String signature,
  							int methodHandle, int actualParameterHandle, Class[] formalParameters)
	{
		this.thisPointer			= thisPointer;
		this.packageName 			= packageName;
		this.methodName 			= methodName;
		this.className 				= className;
		this.signature 				= signature;
		this.methodHandle			= methodHandle;
		this.actualParameterHandle	= actualParameterHandle;
		this.formalParameters		= formalParameters;
	}

	/**
	 * Public constructor to create actions that can be inserted.
	 * TODO: should this be a separate class, or perhaps specified using a language construct? Anyway, more of
	 *       the work done here can be done at compile time.
	 */
	public ConcreteAction (Object thisPointer, String fullMethName, Object[] actualParameters)
	{
	    try
	    {
			this.actualParameters = actualParameters;
			this.thisPointer = thisPointer;
			int i=fullMethName.indexOf("(");
			String ps = fullMethName.substring(i+1,fullMethName.length()-1);
			fullMethName = fullMethName.substring(0,i);
			i=fullMethName.lastIndexOf(".");
			methodName = fullMethName.substring(i+1);
			fullMethName = fullMethName.substring(0,i);
			i=fullMethName.lastIndexOf(".");
			if (i > 0)
			{
				packageName = fullMethName.substring(0,i);
				className = fullMethName.substring(i+1);
			}
			StringTokenizer t = new StringTokenizer(ps,",");
			StringBuffer sigBuffer = new StringBuffer ();
			sigBuffer.append ('(');
			formalParameters = new Class[t.countTokens()];
			i=0;
			while (t.hasMoreTokens())
			{
				formalParameters[i] = Types.stringToClass(t.nextToken().trim());
				sigBuffer.append (Types.classToString (formalParameters[i]));
				i++;
			}

			sigBuffer.append (')');
			signature = sigBuffer.toString ();
		}
		catch(Exception e)
		{
			System.out.println("<Polymer> Error creating action object:");
			e.printStackTrace();
			System.exit(1);
		}
	}

	public String getPackageName()
	{
		return packageName;
	}

	public String getClassName()
	{
		return className;
	}

	public String getMethodName()
	{
		return methodName;
	}

	public String getSignature()
	{
		return signature;
	}

  	public Object getThisPointer ()
  	{
  		return thisPointer;
  	}

	public int getParameterCount ()
	{
		// This uses formalParameters to avoid having to load the param array
		return formalParameters.length;
	}

	public Object getActualParameter (int index)
	{
		ensureParametersLoaded ();
		return actualParameters[index];
	}

	public Object[] getActualParameters ()
	{
		ensureParametersLoaded ();
		return actualParameters;
	}

	private final void ensureParametersLoaded ()
	{
		if (actualParameters == null)
		{
			actualParameters = loadParameters (methodHandle, actualParameterHandle);
		}
	}

	private final native static Object[] loadParameters (int methodHandle, int actualParameterHandle);

	public Class getFormalParameter (int index)
	{
		ensureParametersLoaded ();
		return formalParameters[index];
	}

  	public String toString ()
  	{
    	if (formalParameters != null)
    	{
      		String s = packageName + "." + className + "." + methodName + "(";
      		for (int i = 0; i < formalParameters.length; i++)
      			s += formalParameters[i].toString () + ",";
      		s += ")";
      		return s;
    	}
    	else
    	{
    		return packageName + "." + className + "." + methodName + " " + signature;
    	}
  	}

  	public String toSimpleString()
  	{
    	return packageName + "." + className + "." + methodName;
  	}
}

