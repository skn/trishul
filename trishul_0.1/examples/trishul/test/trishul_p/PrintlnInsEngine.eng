package trishul.test.trishul_p;
import trishul_p.*;

public class PrintlnInsEngine
extends Engine
{
	boolean first = true;

	public Order query (Action a)
	{
		aswitch(a)
		{
		case <* *.*.println(..)>: if (first) return new InsertOrder(
						new ConcreteAction (this, "trishul.test.trishul_p.PrintlnInsEngine.test(java.lang.String,int)",
						new Object[] {"Inserted string", new Integer (9)}), this);
			break;
		}
		return null;
	}

	private double test (String s, int i)
	{
		System.out.print (s);
		System.out.print (": ");
		System.out.print (i);
		System.out.print ("\n");
		first = false;
		return 3.14159;
	}

	public void handleResult (Action action, Order s, Object result, boolean wasExnThn)
	{
		System.out.print ("HANDLE " + (wasExnThn ? "EXCEPTION" : "NORMAL") + " -> " + result + "\n");
	}


}
