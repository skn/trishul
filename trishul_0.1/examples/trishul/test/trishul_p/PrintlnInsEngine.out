Inserted string: 9
HANDLE NORMAL -> 3.14159
666
NOT ALLOWED
Orignal string
